design/object-store-user.md

```diff
@@ -0,0 +1,131 @@
+### Background
+This proposal is a necessary part of another enhancement to [dynamically provision S3 buckets](https://github.com/rook/rook/issues/2480). 
+
+With the merge of [2076](https://github.com/rook/rook/pull/2076) Rook now supports an object user CRD and its generated secret containing the AWS key pairs. To create a new object store user, a developer creates a simple _CephObjectStoreUser_ yaml file defining the object store and the user’s display name. After applying this file a CephObjectStoreUser CRD is created in the object store’s namespace, and the S3 object user is created.
+
+Some issues with this approach are:
+* the developer needs to know the name and namespace of the object store service, both to create the object store user yaml file, and to know, in advance, the name of the generated secret[1]. 
+* the object store user needs to reside in the same namespace as the target object store.
+
+<sup>[1]  It is desirable that the app developer be able to produce the pod yaml before creation of the object store user and its generated secret, and, thus, the secret’s name must be predictable since it is consumed by the pod. Today, the secret name consists of “rook-ceph-object-user-” + _objectStore’s metadata.Name_ + _CephObjectStoreUser’s metadata.Name_, and is assumed to live in the object store’s namespace. This enhancement proposes to remove the object store name from the secret’s name. Since object store user names must be unique within a namespace, a secret name derived from the CephObjectStoreUser’s `metadata.Name` is also unique, thus the object store name is not required. This supports an advantage of using a storage class (see below), namely, that the developer does not need to know the name (or namespace) of the object store.</sup>
+
+## Proposed Enhancement
+1. Replace the `store` parameter with `storageClass` which references an object store service. Note: this change is **not** backwards compatible, and this proposal does not include a deprecation period due to the assumption that there are few users of the CephObjectStoreUser CRD.
```

Comment:
At this point, we can't make breaking changes to the CRDs since they are declared stable. Instead of replacing the setting, what if we have both options available? For example:
- `store`: If specified, the user and secret would be generated in the same namespace as the object store and user CRD (same as today). Admins would use this option if they want to generate object user creds for their own usage.
- `storageClass`: If specified, the secret would be generated according to the storage class settings. Non-admins would use this option since they wouldn't have access to the rook namespace.

Reply:
Ok. Adding `storageClass` and keeping `store` was part of the original design. The thought on removing support for `store` was to make bucket creation code and testing a bit simpler -- with the belief that there are few users of the User CRD. We can definitely go back to our original proposal and continue to support `store`.

Reply:
+1. I also agree with your _kick-the-tires_ use case.